Add an I203 rule for python 3 incompatible imports.

lucaswiman opened the pull request:
@adamchainz This pull request adds a new rule `I203`, which contains a set of python-3 incompatible imports together with suggested replacements (or `six` shims, if available).  I had previously planned to include these rules as part of a step-by-step python 3 upgrade guide, but never got around to it. 

The work I did put into that project consisted of compiling the most complete list of removals and moves from python 2 to python 3 in existence<sup>*</sup>. I did this by painstakingly reading all the relevant deprecation PEPs, extracting moves from the `six` documentation, and some manual testing on a large (~700,000 lines) Python 2 code base. The result was [this file](https://github.com/lucaswiman/legacy-python-upgrade-guide/blob/2f62591443df5bc0c139e54e9d342d120a3a091c/imports/tox.ini), which was generated by [this script](https://github.com/lucaswiman/legacy-python-upgrade-guide/blob/2f62591443df5bc0c139e54e9d342d120a3a091c/imports/build_banned_imports.py).

## TODO
- [x] Update documentation. I wanted to make sure you're interested in accepting this change before I wrote documentation for it. :-)

<hr>

<sup>*</sup> That I could find, after a fair amount of Googling. If there's already a list like this, or I missed some changes, I'd love to know.

adamchainz commented:
Nice! I think it might be better to include this list under I202 though, maybe by adding a magic line to `banned-modules` like `{python2to3}`. Then we don't need a separate rule or logic for it, and users can combine it with their own extra banned modules related to their 2 to 3 upgrades.

lucaswiman commented:
I thought the separate rule is desirable because:
- It's on by default, which I think is preferable. Linters are usually maximally strict by default, and flake8 already has configuration to turn off specific rules.
- It has separate logic, so the error messages can be more informative and less verbose. There is some code duplication I could factor out of the two rule methods so they can share as much logic as possible.

That said, it's a judgement call, and this is your project, so I'm fine with doing that if you'd still prefer it.

lucaswiman commented:
Updated in 9cd5d4c25a333b9aa9dc28837e4d8b1042cf40d8.

lucaswiman commented:
Alright, @adamchainz I think I've addressed your comments, and I updated the README. Please let me know if there are additional changes you'd like me to make.

adamchainz commented on flake8_tidy_imports.py:
```diff
@@ -42,6 +42,9 @@ def parse_options(cls, options):
                  if line.strip()]
         cls.banned_modules = {}
         for line in lines:
+            if line.strip() == '{python2to3}':
```
lines are already stripped, no need to restrip them

adamchainz commented on flake8_tidy_imports.py:
```diff
@@ -7,7 +7,7 @@
 __author__ = 'Adam Johnson'
 __email__ = 'me@adamj.eu'
-__version__ = '1.0.6'
+__version__ = '1.1.0'
```
don't bump version number, that's done in release commit

adamchainz commented on flake8_tidy_imports.py:
```diff
@@ -132,3 +135,247 @@ def rule_I201(self, node):
+        'SocketServer': 'moved in Python 3. six.moves.socketserver can be used as a drop-in replacement',
+        'sre': 'moved in Python 3. Use re instead',
+        'statvfs': 'moved in Python 3. Use os.statvfs instead',
+        'string.atof': 'moved in Python 3. float can be used as a drop-in replacement',
```
so one problem with all of this is that we are only detecting import statements, like `from string import atof`. Using `import string` then later `string.atof()` is not detected. Not sure if we should be fixing this in a library called `flake8-tidy-imports`...

lucaswiman replied:
`from` imports are quite common, and some coding standards even require them, so I don't think there's a problem with including them here. I agree that attribute access should be out of scope for this library, so it's not a complete solution, but it's better than nothing.

adamchainz commented:
Released in 1.1.0 : https://pypi.org/project/flake8-tidy-imports/
